### PR TITLE
 Integrate Siddhi for adaptive authentication

### PR DESCRIPTION
**Pull request by madurangasiriwardena**
### Proposed changes in this pull request

- Integrate Siddhi for adaptive authentication

### When should this PR be merged

N/A

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/adaptive-authentication/org.wso2.carbon.identity.adaptive.auth/src/main/java/org/wso2/carbon/identity/adaptive/auth/QueryInterface.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    public QueryInterface(Map<String, SiddhiAppRuntime> siddhiAppRunTimeMap) {
+
+        this.siddhiAppRunTimeMap = siddhiAppRunTimeMap;
+    }
+
+    public List<Event> query(String appName, String query) {
+
+        SiddhiAppRuntime appRunTime = siddhiAppRunTimeMap.get(appName);
+        if (appRunTime == null) {
+            log.info("Siddhi app: " + appName + " is not registered/started.");
```
Comment (pulasthi7) [inline components/adaptive-authentication/org.wso2.carbon.identity.adaptive.auth/src/main/java/org/wso2/carbon/identity/adaptive/auth/QueryInterface.java]: shouldn't this be an error log?

File: components/adaptive-authentication/org.wso2.carbon.identity.adaptive.auth/src/main/java/org/wso2/carbon/identity/adaptive/auth/SiddhiEventPublisher.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        if (streamDefinition == null) {
+            log.error("StreamDefinition for " + streamName + " does not exist. Cannot publish the event.");
+            return;
+        }
+
+        List<Object> eventPayload = getEventPayload(payloadData, streamDefinition);
+        try {
+            inputHandler.send(eventPayload.toArray());
+        } catch (InterruptedException e) {
+            log.error(e);
```
Comment (pulasthi7) [inline components/adaptive-authentication/org.wso2.carbon.identity.adaptive.auth/src/main/java/org/wso2/carbon/identity/adaptive/auth/SiddhiEventPublisher.java]: better to include message with app name, stream name etc

File: components/adaptive-authentication/org.wso2.carbon.identity.adaptive.auth/src/main/java/org/wso2/carbon/identity/adaptive/auth/deployer/SiddhiAppDeployer.java [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+        String siddhiApp;
+        try {
+            siddhiApp = getSiddhiApp(siddhiAppPath);
+            boolean result = AdaptiveDataHolder.getInstance().getSiddhiEngine().deployApp(siddhiApp);
+            if (result) {
+                String siddhiAppName = Utils.getSiddhiAppName(siddhiApp);
+                fileNameToAppName.put(siddhiAppPath.getFileName().toString(), siddhiAppName);
+                log.info("Siddhi App : " + siddhiAppName + " deployed Successfully.");
+                if (log.isDebugEnabled()) {
+                    log.info("Siddhi App deployed Successfully. \n" + siddhiApp);
```
Comment (pulasthi7) [inline components/adaptive-authentication/org.wso2.carbon.identity.adaptive.auth/src/main/java/org/wso2/carbon/identity/adaptive/auth/deployer/SiddhiAppDeployer.java]: debug?

File: components/adaptive-authentication/org.wso2.carbon.identity.adaptive.auth/src/main/java/org/wso2/carbon/identity/adaptive/auth/deployer/SiddhiAppDeployer.java [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+                }
+            } else {
+                log.error("Error while deploying siddhiApp from path: " + siddhiAppPath.toAbsolutePath());
+            }
+        } catch (IOException e) {
+            log.error("Error while deploying siddhiApp from path: " + siddhiAppPath.toAbsolutePath());
```
Comment (pulasthi7) [inline components/adaptive-authentication/org.wso2.carbon.identity.adaptive.auth/src/main/java/org/wso2/carbon/identity/adaptive/auth/deployer/SiddhiAppDeployer.java]: log 'e'

File: components/adaptive-authentication/org.wso2.carbon.identity.adaptive.auth/src/main/java/org/wso2/carbon/identity/adaptive/auth/internal/AdaptiveDataHolder.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+package org.wso2.carbon.identity.adaptive.auth.internal;
+
+import org.wso2.carbon.identity.adaptive.auth.EmbeddedSiddhiEngine;
+import org.wso2.carbon.identity.adaptive.auth.QueryInterface;
+import org.wso2.carbon.identity.adaptive.auth.SiddhiEventPublisher;
+import org.wso2.siddhi.core.SiddhiManager;
+
+public class AdaptiveDataHolder {
+    private SiddhiManager siddhiManager;
```
Comment (pulasthi7) [inline components/adaptive-authentication/org.wso2.carbon.identity.adaptive.auth/src/main/java/org/wso2/carbon/identity/adaptive/auth/internal/AdaptiveDataHolder.java]: leave a new line

File: components/adaptive-authentication/org.wso2.carbon.identity.adaptive.auth/pom.xml [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+                        <Export-Package>
+                            !org.wso2.carbon.identity.adaptive.auth.internal,
+                            org.wso2.carbon.identity.adaptive.auth.*; version="${carbon.identity.package.export.version}"
+                        </Export-Package>
+                    </instructions>
+                </configuration>
+            </plugin>
+        </plugins>
+    </build>
+</project>
```
Comment (mefarazath) [inline components/adaptive-authentication/org.wso2.carbon.identity.adaptive.auth/pom.xml]: new line

File: components/adaptive-authentication/org.wso2.carbon.identity.adaptive.auth/src/main/java/org/wso2/carbon/identity/adaptive/auth/js/PublishEventFunction.java [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+ */
+
+package org.wso2.carbon.identity.adaptive.auth.js;
+
+import java.util.Map;
+
+@FunctionalInterface
+public interface PublishEventFunction {
+
+    void publishEvent(String siddhiAppName, String streamName, Map<String, Object> payloadData);
```
Comment (pulasthi7) [inline components/adaptive-authentication/org.wso2.carbon.identity.adaptive.auth/src/main/java/org/wso2/carbon/identity/adaptive/auth/js/PublishEventFunction.java]: Add doc and class comments (Do for all classes)

File: features/adaptive-authentication/org.wso2.carbon.identity.adaptive.authentication.server.feature/pom.xml [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+                        <goals>
+                            <goal>run</goal>
+                        </goals>
+                    </execution>
+                </executions>
+            </plugin>
+        </plugins>
+    </build>
+
+    <properties>
```
Comment (pulasthi7) [inline features/adaptive-authentication/org.wso2.carbon.identity.adaptive.authentication.server.feature/pom.xml]: Move properties to the repo's root pom

File: pom.xml [hunk 9]
```diff
@@ -1435,6 +1437,29 @@
                 <artifactId>org.wso2.carbon.identity.consent.mgt</artifactId>
                 <version>${project.version}</version>
             </dependency>
+
+
+
```
Comment (pulasthi7) [inline pom.xml]: remove above new lines

File: components/adaptive-authentication/org.wso2.carbon.identity.adaptive.auth/pom.xml (hunk 6 above)
Comment (madurangasiriwardena) [reply components/adaptive-authentication/org.wso2.carbon.identity.adaptive.auth/pom.xml]: done

**Pull request merged**